### PR TITLE
Patch finish task panic

### LINKED ISSUE #1763: Fix tasks pallet finish_task panic
There is a panic in finish_task specifically when a BatchExecuted task is finished due to a submit result call which can lead to finish_task being called twice.

1. reproduce issue in unit test
2. patch bug

### PR DESCRIPTION
**Pull request by peektism**
Closes #1763 

The panic is an edge case which happens when the read task ID is equal to the batch task ID. This causes finish_task to be called twice as a consequence of a single call to `submit_task_result`.

Fixes:
- [x] remove panic from calling finish_task twice as consequence of a single submit_task_result call
- [x] keep the first result submitted to finish_task and ignore the second call to it for the edge case described (because the first result is the consequence of processing the event while the second is a user-submitted result)

Tests:
- [x] unit test showing that TaskShard is cleared as a result of finish_task (unchanged yet expected behavior)
- [x] regression test showing that submit_task_result panics in the edge case described at the top of this description
- [x] regression test showing that the first result is kept in storage and only its associated event is emitted

Follow up discussion/issue: https://github.com/Analog-Labs/timechain/issues/1766

### CODE REVIEW

File: pallets/tasks/src/tests.rs [hunk 1]
```diff
@@ -462,3 +462,51 @@ mod bench_helper {
...
+		// Assign the task to the shard
+		Tasks::assign_task(shard, task_id);
+		assert!(TaskShard::<Test>::contains_key(task_id), "Task should be assigned to a shard");
+		Tasks::finish_task(ETHEREUM, task_id, Ok(()));
+		assert!(!TaskShard::<Test>::contains_key(task_id));
+	});
+}
+
+#[test]
+fn test_regression_finish_task_panic() {
```
Comment (libreamartists) [inline pallets/tasks/src/tests.rs]:
```suggestion
fn test_regression_finish_task_does_not_panic() {
```

File: pallets/tasks/src/tests.rs [hunk 2]
```diff
@@ -462,3 +462,51 @@ mod bench_helper {
...
+#[test]
+fn finish_task_removes_task_shard() {
+	new_test_ext().execute_with(|| {
+		register_gateway(ETHEREUM, 42);
+		let shard = create_shard(ETHEREUM, 3, 1);
+		roll(1);
+		let batch_id = BatchIdCounter::<Test>::get();
+		let task_id = Tasks::create_task(ETHEREUM, Task::SubmitGatewayMessage { batch_id });
+		BatchTaskId::<Test>::insert(batch_id, task_id);
+		// Assign the task to the shard
```
Comment (libreamartists) [inline pallets/tasks/src/tests.rs]:
```suggestion
```

**Comment by peektism** [comment]
Not ready for review, not a full patch. This edge case also needs to solve the discrepancy of choosing whether the first or second result should be stored in TaskOutput. Current code emits two events for each submission but that does not make sense either. I think I would prefer to keep the first result rather than the second for this edge case because that is the result associated with processing events vs the one passed into submit_result is user submitted.

File: pallets/tasks/src/lib.rs [hunk 3]
```diff
@@ -538,9 +538,16 @@ pub mod pallet {
 			task_id
 		}
 
-		fn finish_task(network: NetworkId, task_id: TaskId, result: Result<(), ErrorMsg>) {
+		pub(crate) fn finish_task(
+			network: NetworkId,
+			task_id: TaskId,
+			result: Result<(), ErrorMsg>,
+		) {
+			if TaskOutput::<T>::get(task_id).is_some() {
```
Comment (haider-rs) [inline pallets/tasks/src/lib.rs]: But if a task is re-executed and passed then this is a problem bcz it wont let the task store the Ok() result

Comment (libreamartists) [reply pallets/tasks/src/lib.rs]: It would already not store the Ok() result in the situation you described because there is already an early return in the submit_task_result extrinsic before we get to calling finish_task https://github.com/Analog-Labs/timechain/blob/22fe324d33f232d50e37bb048e2f3f6e36a13de1/pallets/tasks/src/lib.rs#L355

Comment (libreamartists) [reply pallets/tasks/src/lib.rs]: I added this additional return for the edge case where finish_task is called twice as a consequence of the same submit_task_result call in which case IMO we should keep the first result (from process_events) instead of overwriting it with the user submitted result

**Review by agryaznov — APPROVED** [review]

File: pallets/tasks/src/lib.rs (hunk 3 above)
Comment (haider-rs) [reply pallets/tasks/src/lib.rs]: Thanks for pointing out

```suggestion
			if let Some(Ok(())) = TaskOutput::<T>::get(task_id) {
```
otherwise contains key does the trick right?

I also think the previous early return is also probelmatic for restarting tasks but i think the check is there for chronicle members to not again and again submit the same result which belongs to same shard 

i would also like input of @dvc94ch in here.

Comment (haider-rs) [reply pallets/tasks/src/lib.rs]: i mean for now PR looks good its just a normal flow for previous early return.

**Review by haider-rs — APPROVED** [review]

File: pallets/tasks/src/lib.rs (hunk 3 above)
Comment (agryaznov) [reply pallets/tasks/src/lib.rs]: yep contains_key() would be better

Comment (libreamartists) [inline pallets/tasks/src/lib.rs]:
```suggestion
			if TaskOutput::<T>::contains_key(task_id) {
```

Comment (libreamartists) [reply pallets/tasks/src/lib.rs]: opened this issue as a follow up and changed it to contains_key

Comment (libreamartists) [reply pallets/tasks/src/lib.rs]: https://github.com/Analog-Labs/timechain/issues/1766
